lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -1,147 +1,427 @@
+    ctaShape[dim] =
+        std::max(int64_t{1}, labeledShape[dim] / cgaLayout.getOutDimSize(dim));
+  }
+
+  ctaLayout = ensureLayoutNotSmallerThan(ctaLayout, ctaShape);
+  ctaLayout = ensureLayoutNotLargerThan(ctaLayout, ctaShape);
+
+  LinearLayout ret = (ctaLayout * cgaLayout).transposeOuts(outDimNames);
+  for (auto dim : ret.getOutDimNames()) {
+    assert(ret.getOutDimSize(dim) == labeledShape[dim]);
+  }
+  return ret;
+}
+
+LinearLayout blockedToLinearLayout(ArrayRef<int64_t> shape,
+                                   BlockedEncodingAttr blocked) {
+  assert(shape.size() == blocked.getOrder().size());
+
+  int rank = shape.size();
+  MLIRContext *ctx = blocked.getContext();
+  SmallVector<StringAttr> outDimNames = standardOutDimNames(ctx, rank);
+
+  const auto &order = blocked.getOrder();
+  LinearLayout ctaLayout =
+      identityND(S("register"), blocked.getSizePerThread(), order,
+                 outDimNames) *
+      identityND(S("lane"), blocked.getThreadsPerWarp(), order, outDimNames) *
+      identityND(S("warp"), blocked.getWarpsPerCTA(), order, outDimNames);
+
+  return combineCtaCgaWithShape(ctaLayout, blocked.getCTALayout(), shape);
+}
+
+LinearLayout ampereMmaToLinearLayout(ArrayRef<int64_t> shape,
+                                     NvidiaMmaEncodingAttr mma) {
+  int rank = shape.size();
+
+  assert(mma.isAmpere());
+  assert(rank == 2 || rank == 3);
+  assert(mma.getInstrShape().size() == rank);
+  assert((rank == 2 && mma.getInstrShape() == ArrayRef<unsigned>({16, 8})) ||
+         (rank == 3 && mma.getInstrShape() == ArrayRef<unsigned>({1, 16, 8})));
+  for (int i = 0; i < mma.getInstrShape().size(); i++) {
+    assert(shape[i] >= mma.getInstrShape()[i]);
+  }
+
+  MLIRContext *ctx = mma.getContext();
+  SmallVector<StringAttr> dimNames = standardOutDimNames(ctx, rank);
+
+  LinearLayout ctaLayout(
+      {{S("register"), {{1, 0}, {0, 8}}},
+       {S("lane"), {{2, 0}, {4, 0}, {0, 1}, {0, 2}, {0, 4}}}},
+      llvm::to_vector(llvm::reverse(ArrayRef(dimNames).take_back(2))));
+
+  ctaLayout *= identityND(
+      S("warp"), mma.getWarpsPerCTA(),
+      llvm::to_vector(llvm::reverse(llvm::seq<unsigned>(rank))), dimNames);
+
+  return combineCtaCgaWithShape(ctaLayout, mma.getCTALayout(), shape);
+}
+
+LinearLayout hopperMmaToLinearLayout(ArrayRef<int64_t> shape,
+                                     NvidiaMmaEncodingAttr mma) {
+  int rank = shape.size();
+  assert(mma.isHopper());
+  assert(rank == 2);
+
+  // wgmma operates on groups of 4 warps.
+  assert(product(mma.getWarpsPerCTA()) % 4 == 0);
+
+  // Check that it's a known MMA layout.
+  assert(mma.getInstrShape().size() == 3);
+
+  // TODO(jlebar): I'm not sure the names m/k/n are in the right order.  It
+  // doesn't make much sense to me that `k` would affect regBases, since an #mma
+  // layout is for the *output* of the gemm.
+  int m = mma.getInstrShape()[0];
+  int k = mma.getInstrShape()[1];
+  int n = mma.getInstrShape()[2];
```

Comment:
my understanding is that k kind of gives you the type of mma in the layouts described there which is basically linked to the data type:
https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#asynchronous-warpgroup-level-matrix-fragment

The order is definitely what you have

Reply:
Is it? NVIDIA uses m,n,k (and not m,k,n) order to describe their MMA instructions, you can see that in the instruction name.

[This piece of code](https://github.com/triton-lang/triton/blob/a5b32a8718b64786290a030c289c899554364e53/include/triton/Conversion/TritonGPUToLLVM/Utility.h#L707C1-L710C79) seems to imply that something related to M and N is in consecutive indices of the instruciton shape, and not 2 inidices apart?

Reply:
huh you're right I read too fast, it is definitely m, n k here

Reply:
Thank you for the review, @ThomasRaoux, and thanks for the comment, @gflegar.  I've updated the PR and will check this in!